README.md
# level-web-stream

**Read and write to an [`abstract-level`](https://github.com/Level/abstract-level) database using [Web Streams](https://developer.mozilla.org/en-US/docs/Web/API/Streams_API).** Compatible with browsers and Node.js.

> :pushpin: To instead consume data using Node.js streams, see [`level-read-stream`](https://github.com/Level/read-stream). On Node.js 16, `level-read-stream` is ~3 times faster than `level-web-stream` and the performance of Web Streams isn't likely to improve anytime soon. On the other hand, Web Streams are a step towards a standard library for JavaScript (across Node.js, Deno and browsers).

[![level badge][level-badge]](https://github.com/Level/awesome)
[![npm](https://img.shields.io/npm/v/level-web-stream.svg)](https://www.npmjs.com/package/level-web-stream)
[![Node version](https://img.shields.io/node/v/level-web-stream.svg)](https://www.npmjs.com/package/level-web-stream)
[![Test](https://img.shields.io/github/workflow/status/Level/web-stream/Test?label=test)](https://github.com/Level/web-stream/actions/workflows/test.yml)
[![Coverage](https://img.shields.io/codecov/c/github/Level/web-stream?label=&logo=codecov&logoColor=fff)](https://codecov.io/gh/Level/web-stream)
[![Standard](https://img.shields.io/badge/standard-informational?logo=javascript&logoColor=fff)](https://standardjs.com)
[![Common Changelog](https://common-changelog.org/badge.svg)](https://common-changelog.org)
[![Donate](https://img.shields.io/badge/donate-orange?logo=open-collective&logoColor=fff)](https://opencollective.com/level)

## Usage

### Reading

```js
const { EntryStream } = require('level-web-stream')
const { MemoryLevel } = require('memory-level')

const db = new MemoryLevel()

// Write sample data
await db.put('a', '1')
await db.put('b', '2')
await db.put('c', '3')

// Create a ReadableStream
const src = new EntryStream(db, {
  gte: 'b'
})

// Pipe to a stream of choice
const dst = new WritableStream({
  write ([key, value]) {
    console.log('%s: %s', key, value)
  }
})

await src.pipeTo(dst)
```

Yields:

```
b: 2
c: 3
```

To only read keys or values rather than entries:

```js
const { KeyStream, ValueStream } = require('level-web-stream')

await new KeyStream(db).pipeTo(new WritableStream({
  write (key) {
    console.log(key)
  }
}))
```

Note that [`WritableStream`](https://developer.mozilla.org/en-US/docs/Web/API/WritableStream) is a global in browsers. In Node.js it can be imported from [`stream/web`](https://nodejs.org/api/webstreams.html).

### Writing

```js
const { EntryStream, BatchStream } = require('level-web-stream')

// Copy entries from db to db
const src = new EntryStream(db1)
const dst = new BatchStream(db2, { type: 'put' })

await src.pipeTo(dst)
```

## Install

With [npm](https://npmjs.org) do:

```
npm install level-web-stream
```

## API

### `stream = new EntryStream(db[, options])`

Create a [`ReadableStream`](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream) that will yield entries. An entry is an array with two elements: a `key` and `value`. The `db` argument must be an `abstract-level` database. The optional `options` object may contain:

- `highWaterMark` (number): the maximum number of entries to buffer internally before ceasing to read further entries. Default 1000.

Any other options are forwarded to `db.iterator(options)`. The stream wraps that iterator. If you prefer to consume entries with `for await...of` then it's recommended to directly use `db.iterator()`. In either case, most databases will read from a snapshot (thus unaffected by simultaneous writes) as indicated by `db.supports.snapshots`.

### `stream = new KeyStream(db[, options])`

Same as `EntryStream` but yields keys instead of entries, using `db.keys()` instead of `db.iterator()`. If only keys are needed, using `KeyStream` may increase performance because values won't have to be fetched.

### `stream = new ValueStream(db[, options])`

Same as `EntryStream` but yields values instead of entries, using `db.values()` instead of `db.iterator()`. If only values are needed, using `ValueStream` may increase performance because keys won't have to be fetched.

### `stream = new BatchStream(db[, options])`

Create a [`WritableStream`](https://developer.mozilla.org/en-US/docs/Web/API/WritableStream) that takes _operations_ or _entries_, to be written to the database in batches of fixed size using `db.batch()`. If a batch fails the stream will be aborted but previous batches that did succeed will not be reverted.

An _operation_ is an object containing:

- A `key` property (required)
- A `type` property (optional, one of `'put'`, `'del'`)
- A `value` property (required if type is `'put'`, ignored if type is `'del'`)
- Other operation properties accepted by `db.batch()`.

An _entry_ is an array with two elements: a `key` and `value`. This allows piping a readable `EntryStream` into a writable `BatchStream`. Writing `[key, value]` is the same as writing `{ key, value }`.

The `db` argument must be an `abstract-level` database. The optional `options` object may contain:

- `highWaterMark` (number, default 500): the maximum number of operations to buffer internally before committing them to the database with `db.batch()`. No data will be committed until `highWaterMark` is reached or until the stream is closing. Increasing `highWaterMark` can improve throughput at the cost of memory usage and at risk of blocking the JavaScript event loop while `db.batch()` is copying data.
- `type` (string, default `'put'`): default operation `type` if not set on individual operations or entries (which can't set it). Must be `'put'` or `'del'`.

Any other options are forwarded to the `options` argument of `db.batch(operations, options)` thus following the same rules of precedence (of options versus operation properties).

## Contributing

[`Level/web-stream`](https://github.com/Level/web-stream) is an **OPEN Open Source Project**. This means that:

> Individuals making significant and valuable contributions are given commit-access to the project to contribute as they see fit. This project is more like an open wiki than a standard guarded open source project.

See the [Contribution Guide](https://github.com/Level/community/blob/master/CONTRIBUTING.md) for more details.

## License

[MIT](LICENSE)

[level-badge]: https://leveljs.org/img/badge.svg


CHANGELOG.md
# Changelog

## [1.1.0] - 2022-03-25

### Added

- Add writable `BatchStream` ([#2](https://github.com/Level/web-stream/issues/2)) ([`2b892a5`](https://github.com/Level/web-stream/commit/2b892a5)) (Vincent Weevers).

## [1.0.0] - 2022-02-19

_:seedling: Initial release._

[1.1.0]: https://github.com/Level/web-stream/releases/tag/v1.1.0

[1.0.0]: https://github.com/Level/web-stream/releases/tag/v1.0.0


test.js
'use strict'

const test = require('tape')
const { MemoryLevel } = require('memory-level')
const { ReadableStream, WritableStream } = require('./streams')
const { EntryStream, KeyStream, ValueStream, BatchStream } = require('.')

let db
let lastIterator

const data = [
  { key: 'a', value: '1' },
  { key: 'b', value: '2' },
  { key: 'c', value: '3' }
]

test('setup', function (t) {
  db = new MemoryLevel()

  // Keep track of last created iterator for test purposes
  for (const method of ['iterator', 'keys', 'values']) {
    const original = db[method]

    db[method] = function (...args) {
      const it = lastIterator = original.apply(this, args)
      return it
    }
  }

  db.open(function (err) {
    t.error(err, 'no error')
    db.batch(data.map(x => ({ type: 'put', ...x })), function (err) {
      t.error(err, 'no error')
      t.end()
    })
  })
})

test('EntryStream', async function (t) {
  t.same(await concat(new EntryStream(db)), data.map(kv => [kv.key, kv.value]))
})

test('KeyStream', async function (t) {
  t.same(await concat(new KeyStream(db)), data.map(kv => kv.key))
})

test('ValueStream', async function (t) {
  t.same(await concat(new ValueStream(db)), data.map(kv => kv.value))
})

for (const Ctor of [EntryStream, KeyStream, ValueStream]) {
  const name = Ctor.name

  test(name + ': natural end closes iterator', async function (t) {
    const stream = new Ctor(db)
    const order = monitor()
    const items = await concat(stream)

    t.is(items.length, 3, 'got items')
    t.same(order, ['_nextv', '_nextv', '_close'], 'order was ok')
  })

  test(name + ': bubbles up error from iterator.nextv', async function (t) {
    t.plan(2)

    const stream = new Ctor(db)
    const order = monitor()

    lastIterator._nextv = function (size, options, cb) {
      process.nextTick(cb, new Error('nextv test error'))
    }

    try {
      await concat(stream)
    } catch (err) {
      t.is(err.message, 'nextv test error', 'got error')
      t.same(order, ['_close'], 'order was ok')
    }
  })

  test(name + ': cancelation closes iterator', async function (t) {
    t.plan(3)

    const stream = new Ctor(db, { highWaterMark: 1 })
    const order = monitor()

    let count = 0

    try {
      await stream.pipeTo(new WritableStream({
        write (entry) {
          if (++count > 1) throw new Error('test error')
        }
      }))
    } catch (err) {
      t.is(count, 2)
      t.is(err.message, 'test error')
      t.is(order.pop(), '_close', 'iterator was closed')
    }
  })

  test(name + ': signal abort closes iterator', async function (t) {
    const stream = new Ctor(db)
    const order = monitor()
    const controller = new AbortController()

    try {
      await stream.pipeTo(new WritableStream({
        write (item) {
          controller.abort()
        }
      }), { signal: controller.signal })
    } catch (err) {
      t.is(err.name, 'AbortError')
      t.is(order[order.length - 1], '_close', 'iterator was closed')
    }
  })
}

// Done with readable stream tests; subsequent tests use a fresh db
test('teardown', function (t) {
  return db.close()
})

test('BatchStream: basic default operation', async function (t) {
  const { db, batchCalls } = await monkeyBatched()
  const stream = new BatchStream(db)

  await from([{ key: 'a', value: '1' }]).pipeTo(stream)

  t.same(await db.iterator().all(), [['a', '1']])
  t.same(batchCalls, [[[{ key: 'a', value: '1', type: 'put' }], {}]])
})

test('BatchStream: basic put operation', async function (t) {
  const { db, batchCalls } = await monkeyBatched()

  await from([{ type: 'put', key: 'a', value: '1' }]).pipeTo(new BatchStream(db))
  await from([{ key: 'b', value: '2' }]).pipeTo(new BatchStream(db, { type: 'put' }))

  t.same(await db.iterator().all(), [['a', '1'], ['b', '2']])
  t.same(batchCalls, [
    [[{ key: 'a', value: '1', type: 'put' }], {}],
    [[{ key: 'b', value: '2', type: 'put' }], {}]
  ])
})

test('BatchStream: basic del operation', async function (t) {
  const { db, batchCalls } = await monkeyBatched()

  await db.put('a', '1')
  await db.put('b', '2')

  t.same(await db.iterator().all(), [['a', '1'], ['b', '2']])

  await from([{ type: 'del', key: 'a' }]).pipeTo(new BatchStream(db))
  await from([{ key: 'b' }]).pipeTo(new BatchStream(db, { type: 'del' }))

  t.same(await db.iterator().all(), [])
  t.same(batchCalls, [
    [[{ key: 'a', type: 'del' }], {}],
    [[{ key: 'b', type: 'del' }], {}]
  ])
})

test('BatchStream: operation type takes precedence', async function (t) {
  const { db } = await monkeyBatched()

  await db.put('a', '1')
  t.same(await db.values().all(), ['1'])

  await from([{ type: 'put', key: 'a', value: '2' }]).pipeTo(new BatchStream(db, { type: 'del' }))
  t.same(await db.values().all(), ['2'])
})

test('BatchStream: basic operation with custom property', async function (t) {
  const { db, batchCalls } = await monkeyBatched()
  const stream = new BatchStream(db)
  await from([{ key: 'a', value: '1', foo: 123 }]).pipeTo(stream)
  t.same(batchCalls, [[[{ key: 'a', value: '1', foo: 123, type: 'put' }], {}]])
})

test('BatchStream: basic operation with options', async function (t) {
  const { db, batchCalls } = await monkeyBatched()
  const stream = new BatchStream(db, { foo: 123 })
  await from([{ key: 'a', type: 'del' }]).pipeTo(stream)
  t.same(batchCalls, [[[{ key: 'a', type: 'del' }], { foo: 123 }]])
})

test('BatchStream: basic entry', async function (t) {
  const { db, batchCalls } = await monkeyBatched()
  await from([['a', '1']]).pipeTo(new BatchStream(db))

  t.same(await db.iterator().all(), [['a', '1']])
  t.same(batchCalls, [[[{ key: 'a', value: '1', type: 'put' }], {}]])
})

test('BatchStream: basic entry with del type', async function (t) {
  const { db, batchCalls } = await monkeyBatched()

  await db.put('a', '1')
  await db.put('b', '2')

  t.same(await db.iterator().all(), [['a', '1'], ['b', '2']])

  await from([['a', '1']]).pipeTo(new BatchStream(db, { type: 'del' }))
  await from([['b']]).pipeTo(new BatchStream(db, { type: 'del' }))

  t.same(await db.iterator().all(), [])
  t.same(batchCalls, [
    [[{ key: 'a', value: '1', type: 'del' }], {}],
    [[{ key: 'b', value: undefined, type: 'del' }], {}]
  ])
})

test('BatchStream: uses batches of fixed size', async function (t) {
  const { db, batchCalls } = await monkeyBatched()
  const entries = new Array(17).fill().map((_, i) => [String(i), String(i)])

  await from(entries).pipeTo(new BatchStream(db, { highWaterMark: 5 }))

  t.is(batchCalls.length, 4, '4 batches')
  t.same(batchCalls.map(args => args[0].length), [5, 5, 5, 2], 'batch size is fixed')
})

test('BatchStream: flushes remainder on close', async function (t) {
  const { db, batchCalls } = await monkeyBatched()
  const entries = new Array(5).fill().map((_, i) => [String(i), String(i)])

  await from(entries).pipeTo(new BatchStream(db, { highWaterMark: 1e3 }))

  t.is(batchCalls.length, 1, '1 batch')
  t.is(batchCalls[0][0].length, 5, '5 operations')
})

test('BatchStream: does not batch() in parallel', async function (t) {
  t.plan(5)

  const db = new MemoryLevel()
  const originalBatch = db.batch
  const entries = new Array(10).fill().map((_, i) => [String(i), String(i)])

  // Our monkey-patched batch() does not handle deferred open
  await db.open()

  let inflight = 0

  db.batch = async function (...args) {
    t.is(inflight++, 0, 'not parallel')

    await originalBatch.apply(this, args)
    await new Promise((resolve) => setTimeout(resolve, 50))

    inflight = 0
  }

  await from(entries).pipeTo(new BatchStream(db, { highWaterMark: 2 }))
})

// Create a fresh db with a spy on batch()
const monkeyBatched = async function () {
  const db = new MemoryLevel()
  const originalBatch = db.batch
  const batchCalls = []

  db.batch = function (...args) {
    batchCalls.push(args)
    return originalBatch.apply(this, args)
  }

  // Our monkey-patched batch() does not handle deferred open
  await db.open()

  return { db, batchCalls }
}

const monitor = function () {
  const order = []

  ;['_next', '_nextv', '_close'].forEach(function (method) {
    const original = lastIterator[method]

    lastIterator[method] = function () {
      order.push(method)
      original.apply(this, arguments)
    }
  })

  return order
}

const concat = async function (stream) {
  const items = []

  await stream.pipeTo(new WritableStream({
    write (item) {
      items.push(item)
    }
  }))

  return items
}

// Create a readable stream from an array of items
const from = function (items) {
  let position = 0

  return new ReadableStream({
    pull (controller) {
      if (position >= items.length) {
        controller.close()
      } else {
        controller.enqueue(items[position++])
      }
    }
  })
}


index.d.ts
// Assumed to be installed side-by-side, declared as an optional peerDependency.
import {
  AbstractLevel,
  AbstractIteratorOptions,
  AbstractKeyIteratorOptions,
  AbstractValueIteratorOptions,
  AbstractBatchOptions,
  AbstractBatchOperation
} from 'abstract-level'

/**
 * A {@link ReadableStream} that yields entries.
 */
export class EntryStream<K, V, TDatabase = AbstractLevel<any, any, any>> extends ReadableStream<{ key: K, value: V }> {
  /**
   * Create a {@link ReadableStream} that yields entries from {@link db}.
   * @param db Database to read from.
   * @param options Options for the stream and its underlying iterator.
   */
  constructor (db: TDatabase, options?: (LevelReadableStreamOptions & Omit<AbstractIteratorOptions<K, V>, 'keys' | 'values'>) | undefined)

  // TODO: support passing in an iterator so that its implementation-specific options are typed?
  // constructor (iterator: AbstractIterator<TDatabase, K, V>, ...)
}

/**
 * A {@link ReadableStream} that yields keys.
 */
export class KeyStream<K, TDatabase = AbstractLevel<any, any, any>> extends ReadableStream<K> {
  /**
   * Create a {@link ReadableStream} that yields keys from {@link db}.
   * @param db Database to read from.
   * @param options Options for the stream and its underlying iterator.
   */
  constructor (db: TDatabase, options?: (LevelReadableStreamOptions & AbstractKeyIteratorOptions<K>) | undefined)
}

/**
 * A {@link ReadableStream} that yields values.
 */
export class ValueStream<K, V, TDatabase = AbstractLevel<any, any, any>> extends ReadableStream<V> {
  /**
   * Create a {@link ReadableStream} that yields values from {@link db}.
   * @param db Database to read from.
   * @param options Options for the stream and its underlying iterator.
   */
  constructor (db: TDatabase, options?: (LevelReadableStreamOptions & AbstractValueIteratorOptions<K, V>) | undefined)
}

declare interface LevelReadableStreamOptions {
  /**
   * The maximum number of items to buffer internally before ceasing to read further
   * items.
   *
   * @defaultValue `1000`
   */
  highWaterMark?: number | undefined

  /**
   * Limit the amount of data that the underlying iterator will hold in memory.
   *
   * Only supported by [`classic-level`][1] and [`rocks-level`][2], and possibly by
   * similar `abstract-level` implementations that are backed by a database on disk.
   *
   * [1]: https://github.com/Level/classic-level
   * [2]: https://github.com/Level/rocks-level
   */
  highWaterMarkBytes?: number | undefined

  /**
   * Only supported by [`classic-level`][1] and [`rocks-level`][2], and possibly by
   * similar `abstract-level` implementations that are backed by a database on disk.
   *
   * [1]: https://github.com/Level/classic-level
   * [2]: https://github.com/Level/rocks-level
   */
  fillCache?: boolean | undefined
}

/**
 * A {@link WritableStream} that takes _operations_ or _entries_.
 */
export class BatchStream<K, V, TDatabase = AbstractLevel<any, any, any>>
  extends WritableStream<AbstractBatchOperation<TDatabase, K, V> | [K, V]> {
  /**
   * Create a {@link WritableStream} that takes _operations_ or _entries_, to be
   * written to {@link db} in batches of fixed size using `db.batch()`.
   *
   * @param db Database to write to.
   * @param options Options for the stream and `db.batch()`.
   */
  constructor (
    db: TDatabase,
    options?: (BatchStreamOptions & AbstractBatchOptions<K, V>) | undefined
  )
}

/**
 * Stream options for {@link BatchStream}.
 */
declare interface BatchStreamOptions {
  /**
   * The maximum number of operations to buffer internally before
   * committing them to the database with `db.batch()`.
   *
   * @defaultValue `500`
   */
  highWaterMark?: number | undefined

  /**
   * Default operation `type` if not set on individual operations or entries (which can't
   * set it).
   *
   * @defaultValue `'put'`
   */
  type?: 'put' | 'del'
}


index.js
'use strict'

const { ReadableStream, CountQueuingStrategy, WritableStream } = require('./streams')

const kCanceled = Symbol('canceled')
const kIterator = Symbol('iterator')
const kOperations = Symbol('operations')
const kBatchSize = Symbol('batchSize')
const kBatchType = Symbol('batchType')
const kBatchOptions = Symbol('batchOptions')
const kDb = Symbol('db')

class LevelSource {
  constructor (iterator) {
    this[kIterator] = iterator
    this[kCanceled] = false
  }

  /**
   * @param {ReadableStreamDefaultController} controller
   */
  async pull (controller) {
    let items

    try {
      items = await this[kIterator].nextv(controller.desiredSize)
    } catch (err) {
      await this[kIterator].close()
      throw err
    }

    // Nothing to do if cancel() was called
    if (this[kCanceled]) {
      return
    }

    if (items.length === 0) {
      // Reached the natural end of the iterator
      await this[kIterator].close()
      controller.close()
    } else {
      for (const item of items) {
        controller.enqueue(item)
      }
    }
  }

  cancel () {
    this[kCanceled] = true
    return this[kIterator].close()
  }
}

class LevelReadableStream extends ReadableStream {
  constructor (db, method, options) {
    const { highWaterMark, ...rest } = options || {}
    const iterator = db[method](rest)
    const source = new LevelSource(iterator)
    const queueingStrategy = new CountQueuingStrategy({
      highWaterMark: highWaterMark || 1000
    })

    super(source, queueingStrategy)

    // Keep db around to prevent GC
    this[kDb] = db
  }
}

class EntryStream extends LevelReadableStream {
  constructor (db, options) {
    super(db, 'iterator', { ...options, keys: true, values: true })
  }
}

class KeyStream extends LevelReadableStream {
  constructor (db, options) {
    super(db, 'keys', options)
  }
}

class ValueStream extends LevelReadableStream {
  constructor (db, options) {
    super(db, 'values', options)
  }
}

class LevelSink {
  constructor (db, batchSize, batchType, batchOptions) {
    this[kDb] = db
    this[kOperations] = []
    this[kBatchSize] = batchSize
    this[kBatchType] = batchType
    this[kBatchOptions] = batchOptions
  }

  write (operation) {
    if (Array.isArray(operation)) {
      operation = {
        key: operation[0],
        value: operation[1],
        type: this[kBatchType]
      }
    } else if (operation.type == null) {
      operation = Object.assign({}, operation, {
        type: this[kBatchType]
      })
    }

    const length = this[kOperations].push(operation)

    // Flush if we have a full batch
    if (length >= this[kBatchSize]) {
      const operations = this[kOperations].splice(0, length)
      return this[kDb].batch(operations, this[kBatchOptions])
    }
  }

  close () {
    // Flush remainder if any, returning a promise
    if (this[kOperations].length > 0) {
      return this[kDb].batch(this[kOperations], this[kBatchOptions])
    }
  }
}

class BatchStream extends WritableStream {
  constructor (db, options) {
    let { highWaterMark, type, ...batchOptions } = options || {}

    // Note there are two buffers. Unfortunately Web Streams have no _writev() equivalent
    highWaterMark = highWaterMark || 500
    type = type || 'put'

    const sink = new LevelSink(db, highWaterMark, type, batchOptions)
    const queueingStrategy = new CountQueuingStrategy({ highWaterMark })

    super(sink, queueingStrategy)
  }
}

exports.EntryStream = EntryStream
exports.KeyStream = KeyStream
exports.ValueStream = ValueStream
exports.BatchStream = BatchStream
